tmpl/util.go

```diff
@@ -186,27 +244,79 @@ func (f *tmplFuncs) fullyQualified(typePath string) string {
 	return fmt.Sprintf(".%s.%s", pkg, typePath)
 }
 
+// pkgName parses the named file path for a package statement and returns the
+// package name.
+func (f *tmplFuncs) pkgName(path string) (string, error) {
+	// Grab the name from cache if we already have it.
+	cached, ok := f.pkgNameCache[path]
+	if ok {
+		return cached, nil
+	}
+
+	// Parse for the package name and put into cache.
+	var (
+		file *os.File
+		err  error
+	)
+	for _, includeDir := range f.protoPath {
+		file, err = os.Open(filepath.Join(includeDir, path))
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return "", err
+		}
+		defer file.Close()
+		break
+	}
+	if file == nil {
+		return "", fmt.Errorf("not found in PROTO_PATH: %s", path)
+	}
+
+	// *os.File is -- but bufio.Reader isn't -- an io.Seeker, so we end up reading
+	// the whole file into memory anyway.
+	//
+	// TODO(slimsag): adopt a line-by-line approach instead.
```

Comment:
An idea. Buffering is only really needed for doing the initial regex matching. Once you've done that, you just want to seek back a few bytes and read a few bytes.

As far as I can tell, this should work.

Just pass both the buffered file `bufio.NewReader(file)` as an `io.RuneReader`, and the original `*os.File` as `io.ReadSeeker` to `pkgStmt`. You'd just have to be careful to discard the `io.RuneReader` after you're done with it and you begin using the `io.ReadSeeker`.

`*bufio.Reader` doesn't have a `Close` but you'd effectively be doing this:

```
open file
create bufio.NewReader(file)
use *bufio.Reader
close/discard *bufio.Reader
use file
close file
```